src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets

```diff
@@ -569,7 +588,15 @@ Copyright (c) .NET Foundation. All rights reserved.
     Returns="@(_RestoreGraphEntry)">
 
     <!-- Determine the restore output path -->
-    <PropertyGroup Condition=" '$(PackageReferenceCompatibleProjectStyle)' == 'true' OR '$(RestoreProjectStyle)' == 'ProjectJson' ">
+    <PropertyGroup Condition=" '$(PackageReferenceCompatibleProjectStyle)' == 'true' ">
+      <!-- NuGet writes .props and .targets files to the RestoreOutputPath for PackageReference projects.  MSBuild common
+           targets import these via the MSBuildProjectExtensionsPath, so that's what the RestoreOutputPath should be. -->
+      <RestoreOutputPath Condition=" '$(RestoreOutputPath)' == '' " >$(MSBuildProjectExtensionsPath)</RestoreOutputPath>
+    </PropertyGroup>
+    <PropertyGroup Condition=" '$(RestoreProjectStyle)' == 'ProjectJson' ">
+      <!-- For project.json projects, only the assets file cache goes in the RestoreOutputPath.  The assets file and the
+           generated .props and .targets go in the project folder.  So in that case use the BaseIntermediateOutputPath,
+           so that the assets file cache will respect a BaseIntermediateOutputPath set in the project body. -->
       <RestoreOutputPath Condition=" '$(RestoreOutputPath)' == '' " >$(BaseIntermediateOutputPath)</RestoreOutputPath>
```

Comment:
This is now different between VS and commandline for Project.Json?
One cache file goes BaseIntermediate, the other one goes to MsBuildExtensionsOutputPath.

Why not use MSBuildProjectExtensionsPath here as well?

Reply:
I didn't mean to make it inconsistent, so if it is that should be fixed.

But the idea is that for project.json, the generated props and targets go in the project folder and are imported via other means, so the whole reason we needed to switch to MSBuildProjectExtensionsPath doesn't exist.  So we might as well continue to use BaseIntermediateOutputPath so that we don't change where the assets cache goes if BaseIntermediateOutputPath has been overridden.

Reply:
Valid. 

I think I'll just change what we read in VS for PJ then. 

Other than this change, looks fine to me. 

I'll get someone else to review it and merge it to a feature branch, and fix the tests.

Reply:
What is the issue if we change it for project.json as well? at least that way, it will be a consistent behavior across project formats...

Reply:
@jainaashish 
We can do either one. 
The argument would be, no need to risk changing the experience in PJ, because it's not fixing a bug there. 

I personally don't care too much either way.

Reply:
in the time when we're still supporting pj for legacy projects, I'd recommend we keep the code same across formats instead of making it conditional.